diff --git a/toolchain/iOS.cmake b/toolchain/iOS.cmake
index 076a6b9..2e3291a 100644
--- a/toolchain/iOS.cmake
+++ b/toolchain/iOS.cmake
@@ -47,8 +47,8 @@ endif (CMAKE_UNAME)
 
 # Force the compilers to gcc for iOS
 include (CMakeForceCompiler)
-CMAKE_FORCE_C_COMPILER (clang Apple)
-CMAKE_FORCE_CXX_COMPILER (clang++ Apple)
+CMAKE_FORCE_C_COMPILER (/usr/bin/clang Apple)
+CMAKE_FORCE_CXX_COMPILER (/usr/bin/clang++ Apple)
 set(CMAKE_AR ar CACHE FILEPATH "" FORCE)
 
 # Skip the platform compiler checks for cross compiling
